guardian/newsletters-nx
# Dblatcher newsletter data drafts

## What does this change?

 - the `DraftStorage` class now uses `DraftNewsletterData` (inferred by Zod as a  [deep partial](https://github.com/colinhacks/zod#deepPartial) of `NewsletterData`) as the definition of a `Draft` instead of a generic map of string to primitives (which matched the `WizardFormData` type)
 - adds functions for type-safe converting between `WizardFormData` and `DraftNewsletterData` types, with test coverage.
   - primitive values on the `NewsletterData` are converted to the equivalent on `WizardFormData` eg `NewsletterData.identityName` -> `WizardFormData.identityName`
   - values for nested objects in the `NewsletterData` can be collected on `WizardFormData` by using componded key names - eg `NewsletterData.emailEmbed.name` can be collected from / retrieved to `WizardFormData['emailEmbed.name']`
   - date values, are held as strings or Date objects on the `WizardFormData` and converted to Date objects on `DraftNewsletterData`
   - arrays of strings are supported as a data type, but the  `FormSchema` does not support collecting them yet
 - at the points where the wizard backend was using the `DraftStorage`, the transform functions are used to convert between `DraftNewsletterData` and `WizardFormData`
 - cosmetic changes to the "view draft details" page to demonstrate that the better-typed data is available
 - updates the hard coded test drafts to be in the right format.



## How to test
 - Wizard should work as before, but values will only be saved in the draftStorage if they exist on the `NewsletterData` type - IE if you add a new value on a wizard step, it will need to be added to the `newsletterDataSchema` if it should be saved in the `DraftStorage`
 - Breaking changes to the `newsletterDataSchema` should cause the tests to fail


## How can we measure success?

The Wizard remains agnostic on the type of data being collected on the frontend (still uses the generic `WizardFormData` records) and the conversion to the `DraftNewsletterData` is mostly done in the code in the "newsletter-workflow" library. This will allow us to generalise further, making the "state-machine" code generic as to the the type of data it is being provided.

## Have we considered potential risks?

There are some complexities to address in future PR's:
 - If the `NewsletterData` type becomes more complex, we may need to extend the transform functions to support. It will be important to make sure the step coverage remains comprehensive.
 - how to collect string array values - the `SchemaForm` will need to render a custom input component when it encounters a `z.array(z.string())` on the schema.
 - the transformation logic works and has test coverage, but is a bit obscure. It would benefit from a refactor
